gov/agendas/tracker.go

```diff
@@ -0,0 +1,489 @@
+	// Check if voteCounts are needed
+	for idx := range voteInfo.Agendas {
+		agenda := &voteInfo.Agendas[idx]
+		if agenda.Status != statusDefined && agenda.Status != statusStarted {
+			// check the cache
+			counts := tracker.cachedCounts(agenda.ID)
+			if counts == nil {
+				counts = new(voteCount)
+				var err error
+				counts.yes, counts.abstain, counts.no, err = tracker.voteCounter(agenda.ID)
+				if err != nil {
+					log.Errorf("Error counting votes for %s: %v", agenda.ID, err)
+					continue
+				}
+				tracker.cacheVoteCounts(agenda.ID, counts)
+			}
+			for idx := range agenda.Choices {
+				choice := &agenda.Choices[idx]
+				if choice.ID == choiceYes {
+					choice.Count = counts.yes
+				} else if choice.ID == choiceNo {
+					choice.Count = counts.no
+				} else {
+					choice.Count = counts.abstain
+				}
+			}
+		}
+	}
+	tracker.mtx.Lock()
+	defer tracker.mtx.Unlock()
+	tracker.voteInfo = voteInfo
+	tracker.stakeInfo = stakeInfo
+	ringLen := int(tracker.params.BlockUpgradeNumToCheck)
+	for idx := range blocks {
+		tracker.ringIndex = (tracker.ringIndex + 1) % ringLen
+		tracker.blockRing[tracker.ringIndex] = blocks[idx]
+	}
+	tracker.blockVersion = tracker.blockRing[tracker.ringIndex]
+	tracker.stakeVersion = stakeVersion
+	tracker.ringHeight = voteInfo.CurrentHeight
+	tracker.summary = tracker.newVoteSummary()
+}
+
+// Create a new VoteSummary from the currently saved info.
+func (tracker *VoteTracker) newVoteSummary() *VoteSummary {
+	summary := &VoteSummary{
+		Version:         tracker.version,
+		Height:          tracker.voteInfo.CurrentHeight,
+		Hash:            tracker.voteInfo.Hash,
+		VoteVersion:     tracker.version,
+		MinerThreshold:  tracker.minerThreshold,
+		VoterThreshold:  tracker.voterThreshold,
+		RCIBlocks:       tracker.rciBlocks,
+		SVIBlocks:       tracker.sviBlocks,
+		NextRCIHeight:   uint32(tracker.voteInfo.EndHeight + 1),
+		NetworkUpgraded: uint32(tracker.blockVersion) == tracker.version && tracker.stakeVersion == tracker.version,
+		RCIMined:        uint32(tracker.voteInfo.CurrentHeight - tracker.voteInfo.StartHeight + 1),
+	}
+	summary.Agendas = make([]AgendaSummary, len(tracker.voteInfo.Agendas))
+	summary.RCIProgress = float32(summary.RCIMined) / float32(summary.RCIBlocks)
+
+	for idx := range tracker.voteInfo.Agendas {
+		agenda := &tracker.voteInfo.Agendas[idx]
+		agendaSummary := AgendaSummary{
+			Description:   agenda.Description,
+			ID:            agenda.ID,
+			Quorum:        tracker.params.RuleChangeActivationQuorum,
+			PassThreshold: tracker.passThreshold,
+			// LockCount:     tracker.lockCount,
+		}
+		status := agenda.Status
+		agendaSummary.IsLocked = status == statusLocked
+		agendaSummary.IsFailed = status == statusFailed
+		agendaSummary.IsActive = status == statusActive
+		agendaSummary.IsVoting = status == statusStarted
+		agendaSummary.IsDefined = status == statusDefined
+		for idx := range agenda.Choices {
```

Comment:
This is actually fine since `idx` here is shadowing `idx` of the outer loop, but it's slightly alarming to see.  I noticed you used `idy` in other such cases, so I figured that might have been your intention.